### PR TITLE
MarginTrade and CloseWithSwap using signatures

### PR DESCRIPTION
**Pull request by smbsp**
Allows to open margin trade positions and close positions with swap using signatures. marginTradeBySig() and closeWithSwapWithSignature() are the two functions that are created.

### CODE REVIEW

File: contracts/core/State.sol [hunk 1]
```diff
@@ -204,6 +204,31 @@ contract State is Objects, ReentrancyGuard, Ownable {
...
+	bytes32 public constant DOMAIN_TYPEHASH = keccak256("EIP712Domain(string name,uint256 chainId,address verifyingContract)");
+
+	/// @notice The EIP-712 typehash for the MarginTradeOrder struct used by the contract.
+	bytes32 public constant CLOSE_WITH_SWAP_TYPEHASH =
+		keccak256(
+			"CloseWithSwap(bytes32 loanId,address receiver,uint256 swapAmount,bool returnTokenIsCollateral,bytes32 loanDataBytes,uint256 createdTimestamp)"
+		);
+
+	/// @notice The name of this contract.
+	string public constant NAME = "Loan Closing";
```
Comment (ororopickpocket) [inline contracts/core/State.sol:227]: the State.sol contract is not only used by the loan closing contract

File: contracts/modules/LoanClosingsWith.sol [hunk 2]
```diff
@@ -131,6 +141,132 @@ contract LoanClosingsWith is
...
+		/// @dev CLOSE_WITH_SWAP_TYPEHASH
+		bytes32 structHash = _getStructHash(closePosition);
+
+		bytes32 digest = keccak256(abi.encodePacked("\x19\x01", domainSeparator, structHash));
+		address signatory = ecrecover(digest, v, r, s);
+
+		/// @dev Verify address is not null and PK is not null either.
+		require(RSKAddrValidator.checkPKNotZero(signatory), "LoanClosing::closeWithSwapWithSignature: invalid signature");
+
+		require(!executedOrders[digest], "Order already executed");
```
Comment (ororopickpocket) [inline contracts/modules/LoanClosingsWith.sol:191]: shouldn't order execution handling be on the settlement contract?

File: contracts/core/State.sol [hunk 3]
```diff
@@ -204,6 +204,31 @@ contract State is Objects, ReentrancyGuard, Ownable {
 	mapping(address => mapping(address => uint256)) public specialRebates; // Special rate rebates for spesific pair -- if not set, then use the default one
 	bool public pause; //Flag to pause all protocol modules
 
+	/// ClosePosition
+	struct ClosePosition {
```
Comment (ororopickpocket) [inline contracts/core/State.sol:208]: where is the price data?

**Comment by smbsp** [comment]
This PR was required for Limit Orders - Take Profit/Stop Loss functionality. It has been temporarily put on hold. We will proceed with Limit Orders for Swap and Margin trades and hence archiving this.

**Pull request closed without merging**